### PR TITLE
Enable & error on all un-violated warnings

### PR DESCRIPTION
**Pull request by sellout**
This effectively switches warnings from opt-in to opt-out. It also opts out of
all warnings that currently have violations.

### CODE REVIEW

**Review by nuttycom — DISMISSED** [review]
utACK modulo the Windows build failure

**Comment by daira** [comment]
On macOS, why are we using the Xcode clang[++] rather than the pinned clang[++]?

From https://github.com/zcash/zcash/actions/runs/4849468289/jobs/8641476053?pr=6624 :
```
  CC            = /Users/runner/work/zcash/zcash/depends/x86_64-apple-darwin20.6.0/share/../native/bin/ccache
    /Applications/Xcode_13.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang
    -mmacosx-version-min=10.14 -fvisibility=hidden -fvisibility-inlines-hidden
    --sysroot /Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk
  CFLAGS        = -pipe -O3 
  CPPFLAGS      =   -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -Qunused-arguments
    -I/Users/runner/work/zcash/zcash/depends/x86_64-apple-darwin20.6.0/share/../include/
    -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS -DMAC_OSX
  CXX           = /Users/runner/work/zcash/zcash/depends/x86_64-apple-darwin20.6.0/share/../native/bin/ccache
    /Applications/Xcode_13.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++
    -mmacosx-version-min=10.14 -fvisibility=hidden -fvisibility-inlines-hidden -stdlib=libc++
    --sysroot /Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk
    -std=c++17
  CXXFLAGS      =  -g  -Wformat -Wformat-security -Wstack-protector -fstack-protector-all -Wthread-safety-analysis
    -pipe -O3  -fwrapv -fno-strict-aliasing
  LDFLAGS       = -Qunused-arguments -pthread  -Wl,-pie
    -L/Users/runner/work/zcash/zcash/depends/x86_64-apple-darwin20.6.0/share/../lib
    -static-libstdc++ -lc++abi -Wl,-headerpad_max_install_names -Wl,-dead_strip
  ARFLAGS       = cr

[...]
  CXX      libbitcoin_server_a-addrman.o
error: unknown warning option '-Wno-aix-compat' [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wno-deprecated-copy-with-user-provided-copy' [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wno-deprecated-copy-with-user-provided-dtor'; did you mean '-Wno-deprecated-copy-dtor'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wno-reserved-identifier'; did you mean '-Wno-reserved-id-macro'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wno-unused-but-set-variable'; did you mean '-Wno-unused-const-variable'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wbit-int-extension' [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wc++14-attribute-extensions'; did you mean '-Wcxx-attribute-extension'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wc++17-attribute-extensions'; did you mean '-Wcxx-attribute-extension'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wc++20-attribute-extensions'; did you mean '-Wcxx-attribute-extension'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wdelimited-escape-sequence-extension' [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wfuture-attribute-extensions'; did you mean '-Wcxx-attribute-extension'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wgnu-line-marker'; did you mean '-Wgnu-case-range'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wgnu-null-pointer-arithmetic'; did you mean '-Wnull-pointer-arithmetic'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wgnu-pointer-arith'; did you mean '-Wpointer-arith'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wignored-reference-qualifiers'; did you mean '-Wignored-qualifiers'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Winvalid-utf8'; did you mean '-Winvalid-iboutlet'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wpre-c++2b-compat' [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wunreachable-code-fallthrough'; did you mean '-Wunreachable-code-return'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wunreachable-code-generic-assoc' [-Werror,-Wunknown-warning-option]
make[3]: *** [libbitcoin_server_a-addrman.o] Error 1
make[3]: *** Waiting for unfinished jobs....
error: unknown warning option '-Wno-aix-compat' [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wno-deprecated-copy-with-user-provided-copy' [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wno-deprecated-copy-with-user-provided-dtor'; did you mean '-Wno-deprecated-copy-dtor'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wno-reserved-identifier'; did you mean '-Wno-reserved-id-macro'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wno-unused-but-set-variable'; did you mean '-Wno-unused-const-variable'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wbit-int-extension' [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wc++14-attribute-extensions'; did you mean '-Wcxx-attribute-extension'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wc++17-attribute-extensions'; did you mean '-Wcxx-attribute-extension'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wc++20-attribute-extensions'; did you mean '-Wcxx-attribute-extension'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wdelimited-escape-sequence-extension' [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wfuture-attribute-extensions'; did you mean '-Wcxx-attribute-extension'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wgnu-line-marker'; did you mean '-Wgnu-case-range'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wgnu-null-pointer-arithmetic'; did you mean '-Wnull-pointer-arithmetic'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wgnu-pointer-arith'; did you mean '-Wpointer-arith'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wignored-reference-qualifiers'; did you mean '-Wignored-qualifiers'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Winvalid-utf8'; did you mean '-Winvalid-iboutlet'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wpre-c++2b-compat' [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wunreachable-code-fallthrough'; did you mean '-Wunreachable-code-return'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wunreachable-code-generic-assoc' [-Werror,-Wunknown-warning-option]
make[3]: *** [libbitcoin_server_a-sendalert.o] Error 1
error: unknown warning option '-Wno-aix-compat' [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wno-deprecated-copy-with-user-provided-copy' [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wno-deprecated-copy-with-user-provided-dtor'; did you mean '-Wno-deprecated-copy-dtor'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wno-reserved-identifier'; did you mean '-Wno-reserved-id-macro'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wno-unused-but-set-variable'; did you mean '-Wno-unused-const-variable'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wbit-int-extension' [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wc++14-attribute-extensions'; did you mean '-Wcxx-attribute-extension'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wc++17-attribute-extensions'; did you mean '-Wcxx-attribute-extension'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wc++20-attribute-extensions'; did you mean '-Wcxx-attribute-extension'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wdelimited-escape-sequence-extension' [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wfuture-attribute-extensions'; did you mean '-Wcxx-attribute-extension'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wgnu-line-marker'; did you mean '-Wgnu-case-range'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wgnu-null-pointer-arithmetic'; did you mean '-Wnull-pointer-arithmetic'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wgnu-pointer-arith'; did you mean '-Wpointer-arith'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wignored-reference-qualifiers'; did you mean '-Wignored-qualifiers'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Winvalid-utf8'; did you mean '-Winvalid-iboutlet'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wpre-c++2b-compat' [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wunreachable-code-fallthrough'; did you mean '-Wunreachable-code-return'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wunreachable-code-generic-assoc' [-Werror,-Wunknown-warning-option]
make[3]: *** [zcashd-bitcoind.o] Error 1
```

**Comment by sellout** [comment]
> On macOS, why are we using the Xcode clang[++] rather than the pinned clang[++]?

I added `-Wno-unknown-warning-option` to the macOS build, with a comment explaining that it’s because we use the Xcode clang. Explanation as to _why_ Xcode clang probably belongs somewhere else[^1]. Note that depends/packages/native_clang.mk apparently pulls a Clang 15 for macOS (which I assume is unused and is a bit misleading).

[^1]: depends/hosts/darwin.mk just uses `clang` instead of `$(CC)`, which I’m assuming implicitly uses the Xcode one, since it’s just whatever `clang` is on the `PATH` – and maybe this is why I’ve had trouble getting zcash to build on my Nixified Mac (even pretending it’s x86) where the `clang` on the `PATH` is _not_ Xcode’s.

**Comment by sellout** [comment]
Back to draft as I fix a few macOS issues.

**Comment by sellout** [comment]
I think I’ve fixed all the macOS issues, but until #6628 is merged, it’s hard to be certain.

File: src/Makefile.am
Comment (daira) [inline src/Makefile.am:15]: This is fine; it's different from the options just below because we just don't care about AIX ABI compatibility.

File: src/Makefile.am
Comment (daira) [inline src/Makefile.am:20]: `-Wc++98-compat-pedantic` and `-Wc++11-compat-pedantic` are composite options that include a grab-bag of things that are consistent with the "we don’t want to maintain compatibility with older compilers" rationale, and things that are not.

For example: [`-Wc++11-compat-pedantic`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wc-11-compat-pedantic) implies (among other things) [`-Wc++11-compat`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wc-11-compat). The latter category contains warnings about code that is likely to be incorrect in C++11 and later, even though the code might have been correct &mdash;or differently incorrect&mdash; in earlier versions. In most cases the code being warned about is invalid in C++11 and later. So as far as I can see, there's no advantage in disabling the `-Wc++11-compat` diagnostics because we "don’t want to maintain compatibility with older compilers"; quite the opposite. The same rationale applies to [`-Wc++98-compat`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wc-98-compat). (Not to be confused with [`-Wc++98-compat-pedantic`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wc-98-compat-pedantic)!)

[`-Wc++11-compat-pedantic`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wc-11-compat-pedantic) also implies [`-Wpre-c++20-compat-pedantic`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wpre-c-20-compat-pedantic) and [`-Wpre-c++2b-compat-pedantic`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wpre-c-2b-compat-pedantic), which we should not disable for a different reason: we are not using C++20 yet, and so we want to warn about code that looks like it might be using C++20 constructs that are not supported in C++17. (Just selecting `-std=c++17` doesn't necessarily do so, because the warning might be about code that was technically valid but will change behaviour.)

On the other hand, [`-Wpre-c++14-compat-pedantic`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wpre-c-14-compat-pedantic) and [`-Wpre-c++17-compat-pedantic`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wpre-c-17-compat-pedantic) contain diagnostics that we *don't* want (and so it is correct to use `-Wno-pre-c++14-compat-pedantic` and `-Wno-pre-c++17-compat-pedantic`, I think), because they will just stop us from using features of C++14 and C++17 (or earlier) that are perfectly harmless. These options don't have the same problem as `-Wc++11-compat-pedantic` or `-Wc++98-compat-pedantic`.

Frankly, the inconsistency in semantics of very similarly named options is a mess. The "pedantic" options [aren't documented for GCC](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html), so I don't know whether to blame GCC or clang devs for this, but whoever it was really dropped the ball.

Comment (daira) [reply src/Makefile.am:20]: This is what I think we probably want:
```suggestion
# • `-Wno-c++98-compat`: we want to be able to use C++{11,14,17} constructs introduced after C++98
# • `-Wno-pre-c++14-compat-pedantic`: we want to be able to use C++14 constructs
# • `-Wno-pre-c++17-compat-pedantic`: we want to be able to use C++17 constructs
#
# Note that the naming of '-W*compat*' options is inconsistent; check
# <https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html> before
# changing this.
```

File: src/Makefile.am
Comment (daira) [inline src/Makefile.am:106]:
```suggestion
#    exist in any version of Clang (it does not exist on the Clang used by our other builds).
```

**Review by daira — APPROVED** [review]
utACK with suggestions. These may need tweaking so that they actually suppress all C++ warnings that we need to suppress, but I think that starting with the suggested options and then adding more specific ones is the Right Thing.

File: src/Makefile.am
Comment (sellout) [reply src/Makefile.am:20]: In general, these aggregate flags are designed to be explicitly enabled, rather than disabled after `-Weverything`. So, yes, disabling them silences too many warnings[^1]. I think what we should do is re-enable the compatibility flags we want after disabling the ones we don’t. So here’s the pattern:

1. enable everything
2. disable `*-compat-pedantic` for C++ versions < current
3. enable `*-compat-pedantic` for C++ versions >= current

That (ignoring the flags not related to C++ versions) gets us to
```
-Weverything
-Wno-cxx98-compat-pedantic -Wno-c++11-compat-pedantic -Wno-c++14-compat-pedantic
-Wc++17-compat-pedantic -Wc++20-compat-pedantic
```

Personally, I like the consistency of that, rather than fiddling specific flags. We shouldn’t need `-Wc++17-compat-pedantic` at all (since those things should be errors in the version we’re on), but when we upgrade to C++20, we can switch `-Wc++17-compat-pedantic` to `-Wno` and should be good. And when Clang supports C++23, we add `-Wc++23-compat-pedantic` to the end of the list.

`-Wc++11-compat` _is_ an odd one, but I _think_ it’s ok. Its warnings are things that should be errors since we compile for C++17. E.g., “warning: ’_A_’ is a keyword in C++11” – we would never see that warning even if the flag is enabled (unless the keyword in question was removed in C++14/17), because we’d get an error complaining about _A_ being a keyword.

[^1]: It feels like there’s a variance issue here that is complicated by the order of flags on the command-line mattering. Without significant CLI order some other behavior might make more sense. E.g., enabling a superflag should enable all of its subflags (unless they’ve been explicitly disabled); but disabling a superflag should only disable subflags that aren’t enabled by some other superflag.

Comment (sellout) [reply src/Makefile.am]: I’ve made the change I suggested along with the move to configure.ac – but open to more discussion here.

**Comment by sellout** [comment]
@str4d I moved the warnings to configure.ac.

**Review by daira — COMMENTED** [review]
[ignore]

File: configure.ac
Comment (daira) [inline configure.ac:226]:
```suggestion
                    [Treat many compiler warnings as errors (default is yes)])],
```

File: configure.ac
Comment (daira) [inline configure.ac:358]: We really should make an effort to fix the warnings in [`-Wc++20-compat`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wc-20-compat), because those are blockers for moving to C++20.

Comment (daira) [reply configure.ac:358]: Instead of `-Wno-foo`, what about using `-Wno-error=foo`? That way you still get the warning but it doesn't break the build.

There might be certain warnings that are very noisy, or we don't agree they indicate a warnable issue, that we want to use `-Wno-foo` for. However, I think `-Wno-error=foo` should be the default.

Comment (sellout) [reply configure.ac:358]: Hunh, odd … I saw your first comment here, but not the second in the thread (so I resolved this convo without having read it all).

I think the short answer to `-Wno-error=foo` is “maybe later”. With this PR, that would result in 10s of thousands of warnings (in part because Clang repeats all the warnings from headers every time they get included[^1]). The next PR cleans up most of the flags, but I’m not sure what percentage of the messages that eliminates. However, the flags that are left behind are the more troublesome ones and reporting warnings that _shouldn’t_ be fixed alongside other changes isn’t particularly helpful (especially when you have actual warnings/errors buried amongst the pre-existing noise).

There are a couple of other options – the best is just to get all the warnings turned into errors 😁 but barring that happening quickly, we could
1. silence the remaining warnings on a per-file basis, rather than globally – this prevents contributors from introducing warnings in files that didn’t have that particular warning before, and we can do finer-grained incremental fixing of warnings file-by-file instead of flag-by-flag or
2. have a separate linter that runs the build with `-Wno-error` and `DISABLED_WARNING_CXXFLAGS=""` – this will be a noisy linter that fails for a while, but longer-term I think this is a good approach for warnings that we consider “style warnings”.

Approach 1 is something I’ve done before along with a shrinking whitelist of files that were allowed to have warnings disabled (but I don’t think we need that part – our reviews should catch anyone adding `_Pragma("GCC diagnostic ignored \"-Wfoo\"")`).

Approach 2 is nice because inline warnings can often be missed when the build is successful overall. Seeing a failed check on a PR (even if not a required check) is more likely to get noticed.

[^1]: I wrote a little command-line tool, `dedup`, that replaces blocks of duplicated lines with a reference back to their first occurrence. `make | dedup` is quite helpful for this kind of thing.

File: configure.ac
Comment (daira) [inline configure.ac:780]:
```suggestion
      # • `-Wno-return-std-move-in-c++11`: This causes failure in CI if treated as an error,
      #    despite not being documented to exist in any version of Clang (it does not exist on the
      #    Clang used by our other builds).
```

**Review by daira — DISMISSED** [review]
utACK with suggestions.

After this is merged, I have a PR that will re-enable `-Wcast-align` (#6631).

**Comment by sellout** [comment]
@daira Great – `-Wcast-align` is one I didn't get to. But I do already have commits that re-enable about 80 of the warnings that are silenced here, so hold off on any other changes until I open that PR.

**Review by daira — APPROVED** [review]
utACK 94f13157fb407792b7f856fedd9786be770d127a

**Comment by daira** [comment]
#6635 should fix the macOS build problem in CI.

**Review by nuttycom — APPROVED** [review]
utACK

**Pull request merged**
